### DISCUSSION TITLE
Inquire about the development of the time sheet end time and project deadline notification plug-in

### DISCUSSION THREAD

**Discussion by seleondevel** [kimai/kimai#4734] [Plugins]
Hello.
I would like to create a plug-in in kimai that will be notified when the timesheet ends or when the project schedule ends.
How can I call a notification if I receive a time sheet notification event or a project schedule notification event?

**Comment by kevinpapst** [comment]
There is nothing like that:

> project schedule notification event

Check one of the Timesheet*Event classes in https://github.com/kimai/kimai/tree/main/src/Event, e.g.
https://github.com/kimai/kimai/blob/main/src/Event/TimesheetStopPostEvent.php

> How can I call a notification 

I don't know what that means. Also please share your code and then we can talk about the next steps.

**Reply by seleondevel** [reply]
Even if I install kimai and turn on the browser notification function in my profile, I don't get a notification message, in what case does kimai get a notification?

**Reply by kevinpapst** [reply]
There are no OS notifications triggered yet.

**Reply by seleondevel** [reply]
What do each of the events below mean?
TimesheetStopPreEvent
TimesheetStopPostEvent

Implemented using TimesheetStopPostEvent as below, but event subscribers do not work even when timesheet ends. The contents of the service.yaml code are as follows.

Even if you change it to another event, register or modify the timesheet through kimai, it will not work.

I uploaded a compressed file of the code below.

What should I do?

**Reply by seleondevel** [reply]
[plugin.zip](https://github.com/kimai/kimai/files/14833074/plugin.zip)

**Reply by seleondevel** [reply]
$this->dispatcher->dispatch(new TimesheetStopPreEvent($timesheet));

$this->dispatcher->dispatch(new TimesheetStopPostEvent($timesheet));

I am not receiving this event from EventSubscriber, what should I do?

    public function onTimesheetStopPreEvent(TimesheetStopPreEvent $event)
    {
        $this->logger->info('I\'m logging onTimesheetStopPreEvent');
    }

    public static function getSubscribedEvents()
    {
        return [        
            TimesheetStopPreEvent::class => 'onTimesheetStopPreEvent',
        ];
    }

**Reply by seleondevel** [reply]
Some events are not received during Timesheet*Event events. Can I check it?

And is ProjectStopEvent not implemented?

**Reply by seleondevel** [reply]
public static function getSubscribedEvents()
{
return [
RequestEvent::class => 'onKernelRequest',
ProjectCreateEvent::class => 'onProjectCreateEvent',
TimesheetCreatePostEvent::class => 'onTimesheetCreatePostEvent',
TimesheetCreatePreEvent::class => 'onTimesheetCreatePreEvent',
TimesheetDeleteMultiplePreEvent::class => 'onTimesheetDeleteMultiplePreEvent',
TimesheetDeletePreEvent::class => 'onTimesheetDeletePreEvent',
TimesheetDuplicatePostEvent::class => 'onTimesheetDuplicatePostEvent',
TimesheetDuplicatePreEvent::class => 'onTimesheetDuplicatePreEvent',
TimesheetMetaDefinitionEvent::class => 'onTimesheetMetaDefinitionEvent',
TimesheetMetaDisplayEvent::class => 'onTimesheetMetaDisplayEvent',
TimesheetRestartPostEvent::class => 'onTimesheetRestartPostEvent',
TimesheetRestartPreEvent::class => 'onTimesheetRestartPreEvent',
TimesheetStopPostEvent::class => 'onTimesheetStopPostEvent',
TimesheetStopPreEvent::class => 'onTimesheetStopPreEvent',
TimesheetUpdateMultiplePostEvent::class => 'onTimesheetUpdateMultiplePostEvent',
TimesheetUpdateMultiplePreEvent::class => 'onTimesheetUpdateMultiplePreEvent',
TimesheetUpdatePostEvent::class => 'onTimesheetUpdatePostEvent',
TimesheetUpdatePreEvent::class => 'onTimesheetUpdatePreEvent',
];
}

Is it possible to receive all Timesheet*Event through EventSubscriber?

**Reply by kevinpapst** [reply]
> Is it possible to receive all Timesheet*Event through EventSubscriber?

Yes, why not? You just need a way to trigger them. 

> And is ProjectStopEvent not implemented?

There is no ProjectStopEvent. What should that be? A project has an end date and it can be hidden. But it cannot be "stopped".

**Reply by seleondevel** [reply]
When does an onTimesheetStopPostEvent occur?

**Reply by kevinpapst** [reply]
https://github.com/kimai/kimai/blob/main/src/Timesheet/TimesheetService.php#L218

And this method is called when the stop button is clicked. also when you start a new record, while an old is still running.

**Reply by seleondevel** [reply]
Where is the Stop button? Which button on the kimai homepage should I click to call?

**Reply by seleondevel** [reply]
And, what function should I call to display the notification message in the method of receiving the event in the EventSubscriber class? If the notification function is turned on in the profile on kimai, I would like to send a notification message to the browser, so please let me know which function should be called.

**Reply by kevinpapst** [reply]
>  I would like to send a notification message to the browser, so please let me know which function should be called.

I love the idea, but there is currently no pre-made solution to that.
There are some preparations in the frontend:
https://github.com/kimai/kimai/blob/main/assets/js/plugins/KimaiNotification.js#L59

And it is used on the profile page: 
https://github.com/kimai/kimai/blob/main/templates/user/layout.html.twig

But there needs to be an abstraction layer created, that receives notifications, stores them and injects them in the layout.
In the layout we need to fetch and render them .. kind of, I haven't investigated all possible solutions yet.

**Comment by seleondevel** [comment]
Thank you. As you told me, the timesheet
I checked that the OnTimesheetStopPostEvent event was called when I clicked the Stop button, and I also checked the mail transfer.

When receiving an event using the onProjectCreatePostEvent method after project creation, the ProjectCreatePostEvent function looked up the project information as follows,

public function onProjectCreatePostEvent(ProjectCreatePostEvent $event){

dd($event->getProject()->getName());
}

The project entity does not see the information of the person who created the project.
Where can I get the name of the person who created the project, e-mail address, etc.? Should I access it separately with the User class? I want to send a notification email after creating the project, so please advise.

There is no document organized for the event, so it's too hard to find

**Reply by seleondevel** [reply]
public function onProjectCreatePostEvent(ProjectCreatePostEvent $event){

        if (null === ($user = $this->getUser())) {
            return;
        }

        $this->logger->info(sprintf('ProjectCreatePostEvent!!'));
               
        //dd($this->getUser()->getEmail());
        
        $email = (new Email())
           ->from('admin@xxx.com')
            ->to($this->getUser()->getEmail())           
            ->subject('A project has been created on kimai.")
            ->text('The kimai project was successfully created.");
 
        try {
            $this->mailer->send($email);
            $this->logger->info(sprintf('mail send success!!'));
        } catch (TransportExceptionInterface $e) {
            // some error prevented the email sending; display an
            // error message or try to resend the message
            $this->logger->info(sprintf($e->__toString()));
        }        
    }

 protected function getUser(): ?User
    {
        if (null === $token = $this->tokenStorage->getToken()) {
            return null;
        }

        if (!\is_object($user = $token->getUser()) || !($user instanceof User)) {
            // e.g. anonymous authentication
            return null;
        }

        return $user;
    }

Through $token->getUser() in TokenStorage Interface, we have verified that user information is imported and mail is sent separately
Unable to save changes, the pop-up window continues to pop up.
Why does this pop-up keep popping up when I erase the cache?
How can I resolve this error?

**Reply by seleondevel** [reply]
![스크린샷 2024-04-08 173230](https://github.com/kimai/kimai/assets/165639563/2f0ae652-c493-4b70-8dcc-cd4a192d7736)

**Reply by seleondevel** [reply]
When does this error appear?

**Reply by seleondevel** [reply]
[php] CRITICAL: Uncaught Exception: Symfony\Component\VarDumper\Cloner\Data objects are immutable. {"exception":"[object] (BadMethodCallException(code: 0): Symfony\\Component\\VarDumper\\Cloner\\Data objects are immutable. at /masang/kimai/vendor/symfony/var-dumper/Cloner/Data.php:155)"} []

**Reply by seleondevel** [reply]
Oh. I accidentally changed the eventscriber file name, and it was an error that occurred when it was uploaded in duplicate. This error has been resolved.
